pkg/replication/replication.go

```diff
@@ -278,8 +286,37 @@ func (r replication) updateOne(node types.NodeName, done chan<- types.NodeName,
 		)
 	}
 
-	r.ensureInReality(node, quitCh, nodeLogger, targetSHA)
-	r.ensureHealthy(node, done, quitCh, nodeLogger, aggregateHealth)
+	completedChan := make(chan struct{})
+	updateOneQuit := make(chan struct{})
+	defer close(updateOneQuit)
+
+	go func() {
+		defer close(completedChan)
+		r.ensureInReality(node, updateOneQuit, nodeLogger, targetSHA)
+		r.ensureHealthy(node, updateOneQuit, nodeLogger, aggregateHealth)
+	}()
+
+	if r.timeout == NoTimeout {
+		select {
+		case <-completedChan:
+		case <-quitCh:
+			return
+		}
+	} else {
+		// Wait until either completion, timeout, or a quit
+		select {
+		case <-completedChan:
+		case <-time.After(r.timeout):
+			r.timedOutReplications = append(r.timedOutReplications, node)
```

Comment:
`r.timedOutReplications` is shared state - I'm pretty sure this line needs a mutex around it